components/core/src/clp/ffi/ir_stream/Deserializer.cpp

```diff
@@ -0,0 +1,704 @@
+ * @param reader
+ * @param tag The tag of the value.
+ * @param node_id The node ID of the value.
+ * @param node_id_value_pairs Returns the ID-value pair constructed by the deserialized value.
+ * @return IRErrorCode::IRErrorCode_Success on success.
+ * @return IRErrorCode indicating relevant errors.
+ */
+[[nodiscard]] auto deserialize_value_and_insert_to_node_id_value_pairs(
+        ReaderInterface& reader,
+        encoded_tag_t tag,
+        SchemaTreeNode::id_t node_id,
+        KeyValuePairLogEvent::NodeIdValuePairs& node_id_value_pairs
+) -> IRErrorCode;
+
+/**
+ * Deserializes an encoded text AST and pushes the result into the node_id_value_pairs.
+ * @tparam encoded_variable_t
+ * @param reader
+ * @param node_id The node ID of the value.
+ * @param node_id_value_pairs Returns the ID-value pair constructed by the deserialized encoded text
+ * AST.
+ * @return IRErrorCode::IRErrorCode_Success on success.
+ * @return IRErrorCode indicating relevant errors.
+ */
+template <typename encoded_variable_t>
+requires(std::is_same_v<ir::four_byte_encoded_variable_t, encoded_variable_t>
+         || std::is_same_v<ir::eight_byte_encoded_variable_t, encoded_variable_t>)
+[[nodiscard]] auto deserialize_encoded_text_ast_and_insert_to_node_id_value_pairs(
+        ReaderInterface& reader,
+        SchemaTreeNode::id_t node_id,
+        KeyValuePairLogEvent::NodeIdValuePairs& node_id_value_pairs
+) -> IRErrorCode;
+
+/**
+ * Deserializes values and construct ID-value pairs according to the given schema. The number of
+ * values to deserialize is decided by the size of the given schema.
+ * @param reader
+ * @param tag The current tag.
+ * @param schema The schema of the log event.
+ * @param node_id_value_pairs Returns the constructed ID-value pairs.
+ * @return IRErrorCode::IRErrorCode_Success on success.
+ * @return IRErrorCode indicating relevant errors.
+ */
+[[nodiscard]] auto deserialize_value_and_construct_node_id_value_pairs(
+        ReaderInterface& reader,
+        encoded_tag_t tag,
+        std::vector<SchemaTreeNode::id_t> const& schema,
+        KeyValuePairLogEvent::NodeIdValuePairs& node_id_value_pairs
+) -> IRErrorCode;
+
+auto ir_error_code_to_errc(IRErrorCode ir_error_code) -> std::errc {
+    switch (ir_error_code) {
+        case IRErrorCode_Incomplete_IR:
+            return std::errc::result_out_of_range;
+        case IRErrorCode_Corrupted_IR:
+        case IRErrorCode_Decode_Error:
+            return std::errc::protocol_error;
+        case IRErrorCode_Eof:
+            return std::errc::no_message_available;
+        default:
+            return std::errc::not_supported;
+    }
+}
+
+auto is_schema_tree_node_tag(encoded_tag_t tag) -> bool {
+    return (tag & cProtocol::Payload::SchemaTreeNodeMask) == cProtocol::Payload::SchemaTreeNodeMask;
+}
+
+auto schema_tree_node_tag_to_type(encoded_tag_t tag) -> std::optional<SchemaTreeNode::Type> {
+    std::optional<SchemaTreeNode::Type> type;
```

Comment:
I think `SchemaTreeNode::Type` is a trivial type, so could we return it directly from the method without triggering a copy?